### PR TITLE
AutoEdits: move the semi_automated.yml config to on-wiki JSON page

### PR DESCRIPTION
**Pull request by MusikAnimal**
Bug: T262631

### CODE REVIEW

File: src/AppBundle/Helper/AutomatedEditsHelper.php [hunk 1]
```diff
@@ -65,13 +78,42 @@ public function isAutomated(string $summary, Project $project): bool
...
+            return $this->cache->getItem($cacheKey)->get();
+        }
+
+        $title = 'MediaWiki:XTools-AutoEdits.json' . ($useSandbox ? '/sandbox' : '');
+        $ret = json_decode(file_get_contents(
+            $this->metaProject->getUrl() . "w/index.php?action=raw&ctype=application/json&title=$title"
+        ), true);
+
+        if (!$useSandbox) {
+            $cacheItem = $this->cache
```
Comment (DannyS712) [inline src/AppBundle/Helper/AutomatedEditsHelper.php:92]: Is there any way to invalidate this when the page is edited?

File: src/AppBundle/Repository/AutoEditsRepository.php [hunk 2]
```diff
@@ -49,6 +65,19 @@ public function getTools(Project $project, $namespace = 'all'): array
...
+    /**
+     * Overrides Repository::setCache(), and will not call the parent (which sets the cache) if using the sandbox.
+     * @inheritDoc
+     */
+    public function setCache(string $cacheKey, $value, $duration = 'PT20M')
+    {
+        if ($this->useSandbox) {
+            return $value;
+        } else {
+            return parent::setCache($cacheKey, $value, $duration);
```
Comment (DannyS712) [inline src/AppBundle/Repository/AutoEditsRepository.php]: No need to wrap in an `else` block

File: templates/autoEdits/result.html.twig [hunk 3]
```diff
@@ -216,7 +216,7 @@
                     </div>
                 {% endif %}
                 <p style="clear:both">
-                    <a target="_blank" href="https://phabricator.wikimedia.org/maniphest/task/create/?title=Request to add tool to AutoEdits&amp;projects=XTools&amp;description=Project: {{ project.domain }}%0A%0A{{ msg('auto-edits-request-name') }}:%0A%0A{{ msg('auto-edits-request-examples') }}:%0A%0A{{ msg('auto-edits-request-doc') }}:">
+                    <a target="_blank" href="https://meta.wikimedia.org/w/index.php?title=MediaWiki_talk:XTools-AutoEdits.json&action=edit&section=new&preload=MediaWiki_talk:XTools-AutoEdits.json/Preload">
```
Comment (DannyS712) [inline templates/autoEdits/result.html.twig:219]: Is there also a link to some documentation about how it is now on meta?

File: src/AppBundle/Helper/AutomatedEditsHelper.php (hunk 1 above)
Comment (MusikAnimal) [reply src/AppBundle/Helper/AutomatedEditsHelper.php:92]: Not without querying the db or API (probably would need to be the letter) on every request. Not worth it in my opinion, as we're only talking up to a 20 min wait.

File: templates/autoEdits/result.html.twig (hunk 3 above)
Comment (MusikAnimal) [reply templates/autoEdits/result.html.twig:219]: The thought is requests will mostly come from non-engineers, so I didn't want to confuse them with all the documentation. All maintainers need to know is those 4 questions asked in the preload template. Let's talk more on phab so others can chime in; I've got this along with a number of other things I wanted to ask you all about.

File: src/AppBundle/Helper/AutomatedEditsHelper.php [hunk 4]
```diff
@@ -24,13 +27,23 @@ class AutomatedEditsHelper
     /** @var ContainerInterface The service container. */
     private $container;
 
+    /** @var Project $metaProject */
+    private $metaProject;
+
+    /** @var CacheItemPoolInterface The cache. */
```
Comment (DannyS712) [inline src/AppBundle/Helper/AutomatedEditsHelper.php]: "The cache." doesn't really help document anything since the variable is named $cache and the type is clearly a cache

File: src/AppBundle/Helper/AutomatedEditsHelper.php [hunk 5]
```diff
@@ -24,13 +27,23 @@ class AutomatedEditsHelper
     /** @var ContainerInterface The service container. */
     private $container;
 
+    /** @var Project $metaProject */
```
Comment (DannyS712) [inline src/AppBundle/Helper/AutomatedEditsHelper.php]: $metaProject shouldn't be repeated in the documentation

File: src/AppBundle/Helper/AutomatedEditsHelper.php [hunk 6]
```diff
@@ -24,13 +27,23 @@ class AutomatedEditsHelper
...
      * AutomatedEditsHelper constructor.
      * @param ContainerInterface $container
+     * @param Project|null $metaProject Solely used to mock the meta.wikimedia Project in AutomatedEditsTest
      */
-    public function __construct(ContainerInterface $container)
+    public function __construct(ContainerInterface $container, ?Project $metaProject = null)
     {
         $this->container = $container;
+        $this->cache = $container->get('cache.app');
+        $centralAuthProject = $this->container->getParameter('central_auth_project');
```
Comment (DannyS712) [inline src/AppBundle/Helper/AutomatedEditsHelper.php]: Suggest referencing $container directly, like on line 44, instead of $this->container; same applies to the use of $this->container on line 46

File: src/AppBundle/Helper/AutomatedEditsHelper.php (hunk 1 above)
Comment (DannyS712) [reply src/AppBundle/Helper/AutomatedEditsHelper.php:92]: (noting here what I said on IRC) I meant a way to purge the cache from within the xtools interface

File: src/AppBundle/Helper/AutomatedEditsHelper.php [hunk 7]
```diff
@@ -65,13 +72,45 @@ public function isAutomated(string $summary, Project $project): bool
         return (bool)$this->getTool($summary, $project);
     }
 
+    /**
+     * Fetch the config from https://meta.wikimedia.org/wiki/MediaWiki:XTools-AutoEdits.json
+     * @param bool $useSandbox
```
Comment (DannyS712) [inline src/AppBundle/Helper/AutomatedEditsHelper.php]: Please document that $useSandbox results in a different config source - the current documentation suggest that the main XTools-AutoEdits.json is always used

File: src/AppBundle/Repository/AutoEditsRepository.php [hunk 8]
```diff
@@ -20,6 +20,22 @@ class AutoEditsRepository extends UserRepository
...
+    private $tags;
+
+    /**
+     * AutoEditsRepository constructor. Used solely to set $useSandbox (from AutomatedEditsController).
+     * @param bool $useSandbox
+     */
+    public function __construct(bool $useSandbox = false)
+    {
+        $this->useSandbox = $useSandbox;
+        parent::__construct();
```
Comment (DannyS712) [inline src/AppBundle/Repository/AutoEditsRepository.php:35]: Not sure about within this repo, but generally in my experience the parent::__construct() call is put first

File: src/AppBundle/Repository/AutoEditsRepository.php [hunk 9]
```diff
@@ -61,7 +90,7 @@ public function getTools(Project $project, $namespace = 'all'): array
     public function countAutomatedEdits(Project $project, User $user, $namespace = 'all', $start = '', $end = ''): int
     {
         $cacheKey = $this->getCacheKey(func_get_args(), 'user_autoeditcount');
-        if ($this->cache->hasItem($cacheKey)) {
+        if ($this->cache->hasItem($cacheKey) && !$this->useSandbox) {
```
Comment (DannyS712) [inline src/AppBundle/Repository/AutoEditsRepository.php]: here and below, suggest putting the useSandbox check first to avoid checking the cache unnecessarily, and to be consistent with the order in AutomatedEditsHelper.php (or, if checking second is intentional, suggest being consistent with that)

File: src/AppBundle/Repository/AutoEditsRepository.php [hunk 10]
```diff
@@ -278,19 +307,14 @@ public function getAutomatedEdits(
...
      *                  ]
      */
-    public function getToolCounts(
-        Project $project,
-        User $user,
-        $namespace = 'all',
-        $start = '',
-        $end = ''
-    ): array {
+    public function getToolCounts(Project $project, User $user, $namespace = 'all', $start = '', $end = ''): array
```
Comment (DannyS712) [inline src/AppBundle/Repository/AutoEditsRepository.php:310]: personally, I would prefer to have such long method declarations be split onto multiple lines, but either way suggest doing separately from this patch (as far as I can tell this is a whitespace only change, so it should have no impact on the result and just makes the diff a bit harder to review)

File: src/AppBundle/Repository/AutoEditsRepository.php [hunk 11]
```diff
@@ -323,13 +347,12 @@ public function getToolCounts(
      * Get SQL for getting counts of known automated tools used by the given user.
      * @see self::getAutomatedCounts()
      * @param Project $project
-     * @param User $user
```
Comment (DannyS712) [inline src/AppBundle/Repository/AutoEditsRepository.php:326]: If the user parameter is unused, it should be removed, but how is this related to the rest of the patch? Suggest doing separately

File: tests/AppBundle/Helper/AutomatedEditsTest.php [hunk 12]
```diff
@@ -32,7 +32,6 @@ public function setUp(): void
         $client = static::createClient();
         $container = $client->getContainer();
         $this->aeh = new AutomatedEditsHelper($container);
-        $this->cache = $container->get('cache.app');
```
Comment (DannyS712) [inline tests/AppBundle/Helper/AutomatedEditsTest.php:35]: Same as some of my other comments - how is this related to this patch / can it be done separately?

File: src/AppBundle/Helper/AutomatedEditsHelper.php (hunk 4 above)
Comment (MusikAnimal) [reply src/AppBundle/Helper/AutomatedEditsHelper.php]: Hehe I agree. This is actually just to get around PHPCS from yelling at us. The sniffing rules are probably a little too strict. Anyway I expanded it to "The caching interface" :)

File: src/AppBundle/Repository/AutoEditsRepository.php (hunk 9 above)
Comment (MusikAnimal) [reply src/AppBundle/Repository/AutoEditsRepository.php]: In production it makes virtually no difference, since it's in-memory APCu cache, but good practice nonetheless!

File: src/AppBundle/Repository/AutoEditsRepository.php (hunk 11 above)
Comment (MusikAnimal) [reply src/AppBundle/Repository/AutoEditsRepository.php:326]: It's the XTools way, hehe :) By that I mean I largely work alone here, so little null fixes like this get combined with bigger PRs to save me time. Indeed $user is unused, I'll modify the commit message accordingly, same with below.

File: src/AppBundle/Helper/AutomatedEditsHelper.php (hunk 1 above)
Comment (MusikAnimal) [reply src/AppBundle/Helper/AutomatedEditsHelper.php:92]: https://phabricator.wikimedia.org/T209334

**Comment by MusikAnimal** [comment]
Tests. Finally. Pass. I spent more time fixing that than I did on the actual code changes, 10 times over...

Will get this on xtools-dev before merging

**Comment by MusikAnimal** [comment]
Actually I am going to merge... just to sort of commit to it. Follow-up PRs to come as necesary.

**Comment by MusikAnimal** [comment]
Also thank you again for the review :)
